src/ui/reducers/protocolMessages.ts

```diff
@@ -20,47 +20,53 @@ export type RequestSummary = Omit<CommandRequest, "method"> & {
 
 export interface ProtocolMessagesState {
   events: (ProtocolEvent & Recorded)[];
-  requests: RequestSummary[];
-  responses: (CommandResponse & Recorded)[];
-  errors: (CommandResponse & Recorded)[];
+  idToResponseMap: { [id: number]: CommandResponse & Recorded };
+  idToRequestMap: { [id: number]: RequestSummary };
+  idToErrorMap: { [id: number]: CommandResponse & Recorded };
```

Comment:
As someone who did not do this when first writing this class, I understand the temptation, but especially now that we are grouping things by ID anyways, how would you feel about using entity adapters here?

Reply:
I'm not familiar with that term. Looks like it's a Redux thing? Ah, looking at the [Redux docs](https://redux-toolkit.js.org/api/createEntityAdapter) this looks like what we use with Messages.

I don't think I understand the motivation to add that helper (in this case, or in general). Could you explain?

In this case, I think it would actually add more over-head since it _looks like_ it would result in 3 separate (mostly synced?) id arrays (one per adapter)?

P.S. Can't get past the fact that that API defines a method named `upsertOne` 🤨

Reply:
Decided to skip this one for now. Happy to follow up on it though! Also really interested to hear your thoughts about the benefits.

Reply:
Yeah fair. Honestly I think @markerikson would do a better job of selling it than I probably would. I think in this particular case, an entity adapter probably wouldn't get us a ton *yet*, since the operations that we're doing are really simple, but it seems like the kind of thing that might be nice to have as the code base grows and the way that actions update the state gets more complicated. (I'm sure you've probably seen the redux updates that are like:

```
state: {
  ...state,
    [action.payload.topId]: {
      ...action.payload,
      thisOneTrait: action.payload.thisOtherTrait,
   }
}
```

And while I think Immer takes care of a lot of that entity adapters can also be helpful, particularly for batch updates.

Reply:
I'd need a better understanding of what the actual data being stored is to be able to say whether an entity adapter is the right fit here.  I do see the lookup tables being used, but from Brian's comment I'm not sure what their relationship is.